pkg/challenges/challenges.go

```diff
@@ -273,6 +274,50 @@ func URI(ctx context.Context, principal *oidc.IDToken, pubKey crypto.PublicKey,
 	}, nil
 }
 
+func Username(ctx context.Context, principal *oidc.IDToken, pubKey crypto.PublicKey, challenge []byte) (*ChallengeResult, error) {
+	username := principal.Subject
+
+	cfg, ok := config.FromContext(ctx).GetIssuer(principal.Issuer)
+	if !ok {
+		return nil, errors.New("invalid configuration for OIDC ID Token issuer")
+	}
+
+	// The domain in the configuration must match the domain (excluding the subdomain) of the issuer
+	// In order to declare this configuration, a test must have been done to prove ownership
+	// over both the issuer and domain configuration values.
+	// Valid examples:
+	// * domain = https://example.com/users/user1, issuer = https://accounts.example.com
+	// * domain = https://accounts.example.com/users/user1, issuer = https://accounts.example.com
+	// * domain = https://users.example.com/users/user1, issuer = https://accounts.example.com
+	uIssuer, err := url.Parse(cfg.IssuerURL)
+	if err != nil {
+		return nil, err
+	}
+	if err := isDomainAllowed(cfg.SubjectDomain, uIssuer.Hostname()); err != nil {
+		return nil, err
+	}
+
+	// Check the proof - A signature over the OIDC token subject
+	if err := CheckSignature(pubKey, challenge, username); err != nil {
+		return nil, err
+	}
+
+	issuer, err := oauthflow.IssuerFromIDToken(principal, cfg.IssuerClaim)
+	if err != nil {
+		return nil, err
+	}
+
+	emailSubject := fmt.Sprintf("%s@%s", username, cfg.SubjectDomain)
```

Comment:
Is there any validation we want to do on `username`? I worry about passing things like `zack@example.com` as the username for issuer `somethingelse.com`  and some validation code getting confused down the line.

Less nefariously, we might have `zack@example.com` with issuer `example.com` and getting email `zack@example.com@example.com`

Reply:
Great point! Ultimately it'd be up to the validator to be aware of this, and I'd hope that they're checking matching strings rather than parsing the username. I'll add a check that `@` is not present in the username, that seems very reasonable. If for some reason this comes up again later on, we can relax this requirement.

Another option could be to call [`ParseAddress`](https://pkg.go.dev/net/mail#ParseAddress) after constructing the subject, but I don't think this is necessary, as we aren't requiring this to be a routable email.

Reply:
+1 to do the bare minimum check for `@`, but nothing fancier than that